### ISSUE TITLE
0.9.5-nightly-cc54cba with tsm1 ignores wal-dir

### ISSUE THREAD

**Issue by daviesalex** [influxdb/influxdb#4498]
The WAL design results in a very high number of fsync calls, which even with SAS SSDs is difficult to scale. We want to route WAL requests to a small NVMe volume, but it appears that when using tsm1 InfluxDB is creating the specified WAL directory but then not using it.

The WAL files are all all like this:

```
[root@host data]# lsof | grep wal  | head -2
influxd   30764       influxdb    7u      REG              253,4     479453  249036904 /data/influxdb-data/metrics/telegraf/default/4/_02241.wal
influxd   30764       influxdb   29u      REG              253,4       1484  249036843 /data/influxdb-data/metrics/_internal/monitor/3/_00040.wal
```

There are in fact no open file handlers to the WAL dir:
```
[root@host data]# lsof | grep /data/influxdb-wal
[root@host data]#
```

Our config:
```
[root@host telegraf]# grep wal /etc/influxdb/influxdb.conf
  wal-dir = "/data/influxdb-wal"
  max-wal-size = 10485760000
  wal-flush-interval = "30s"
  wal-partition-flush-delay = "2s"
  wal-logging-enabled = true
  wal-compaction-threshold = 0.5
  wal-max-series-size = 1048576
  wal-flush-cold-interval = "5s"
  wal-flush-memory-size-threshold = 5242880
  wal-max-memory-size-threshold = 10485760000
  wal-partition-size-threshold = 1000000000
  wal-ready-series-size = 512000
```

**Comment by otoolep** [comment]
Yeah, known issue.

**Comment by otoolep** [comment]
With the WAL directory.

Our testing also shows that the `tsm1` directory is quite good with regards to disk IO. Perhaps there is something with regards to your data that is bringing out an issue.

**Comment by daviesalex** [comment]
We agree that the tsm1 directory is good, and vastly better than previous iterations, but it is clearly a huge % of the fast-path IO (operations, not bytes).

It seems that as a general principle, separating out the IO from the backing disks (large capacity) from the WAL (tiny capacity, very many fscync's) is going to be a good practice for InfluxDB as it is for most other databases. In our case, we have a small amount of PCIe SSD per node, which we would like to use for WAL - NVMe was basically designed for the WAL use case (massive numbers of IOPs) but I suspect nobody is going to store their entire database on PCIe flash. Is there a big problem storing the WAL files in a different directory as part of the design of tsm1?

Based on internal feedback, we would also wonder if you could have a less durable option for high loads; i.e. an option to not fsync on each write. This would also hugely reduce the number of IOPS that actually block API calls returning. This would basically be similar to MySQL's innodb-flush-log-at-trx-commit parameter.

What is taking us down this path is that we are continuing to see some TCP connections fail between cluster nodes, and enormous variance in API response time. We do see the occasional bursts in IO service time.

**Comment by otoolep** [comment]
@pauldix -- is splitting the WAL into its own directory for 0.9.5 something we want to commit to? It seems reasonable, for splitting write load across partitions (as previous engine versions supported).

I can make the changes if required, or leave it to you.

**Comment by pauldix** [comment]
I can take care of it. We should let users have a different WAL directory than the index files and there's no compelling reason to force them to use the same dir.

**Comment by daviesalex** [comment]
So after running for 24 hours, we have started to get a lot of errors like this:

```
[tcp] 2015/10/20 09:51:37 process write shard error: write shard 4: engine: WAL backed up flushing to index, hit max memory
[tcp] 2015/10/20 09:51:37 process write shard error: write shard 4: engine: WAL backed up flushing to index, hit max memory
```

At exactly the same time as a large burst of those errors, there is 50MB/sec IO (which is a tiny fraction of what our devices can actually handle - observe the 1second output from iostat below; notice wMB/sec and %util):

```
Device:         rrqm/s   wrqm/s     r/s     w/s    rMB/s    wMB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
dm-4              0.00     0.00    0.00    8.00     0.00     0.03     8.00     0.00    0.12    0.00    0.12   0.12   0.10

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           2.86    0.15    0.44    0.00    0.00   96.54

Device:         rrqm/s   wrqm/s     r/s     w/s    rMB/s    wMB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
dm-4              0.00     0.00    0.00  400.00     0.00    49.43   253.10     4.00   10.01    0.00   10.01   0.05   1.90

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           5.72    0.00    0.26    0.00    0.00   94.02

Device:         rrqm/s   wrqm/s     r/s     w/s    rMB/s    wMB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
dm-4              0.00     0.00    0.00   14.00     0.00     0.05     8.00     0.00    0.14    0.00    0.14   0.14   0.20
```
From this we conclude that
* These errors are correlated with IO
* They are non-desirable, as we have plenty of IO capacity available (and >100G RAM unused)

Its not clear to me why its doing this as we have plenty of IO capacity, but this appears to be causing legitimate problems. Our goal is to configure the WAL to flush early and often but not block - and to not start blocking until we have a huge amount of data backlogged in RAM (bonus points for moving the WAL to a smaller, faster, device - appreciate you working on that for 0.9.5!).

If our config below is not achieving what we want, thats because i'm actually pretty confused from reading a mix of code and documentation, and Paul's talk!

```
[root@host ~]# grep wal /etc/influxdb/influxdb.conf
  max-wal-size = 10485760000
  wal-flush-interval = "30s"
  wal-partition-flush-delay = "2s"
  wal-dir = "/data/influxdb-wal"
  wal-logging-enabled = true
  wal-compaction-threshold = 0.5
  wal-max-series-size = 1048576
  wal-flush-cold-interval = "5s"
  wal-flush-memory-size-threshold = 5242880
  wal-max-memory-size-threshold = 10485760000
  wal-partition-size-threshold = 1000000000
  wal-ready-series-size = 512000
[root@carf-testmetricsl6 ~]#
```

**Comment by pauldix** [comment]
@daviesalex what's the distribution of time on the data you're writing? Are you always writing new (current timestamp) data? How frequently are writes coming into the server?

**Comment by pauldix** [comment]
@daviesalex also, can you link to a gist that has an `ls -lah` of the directory for the highest number shard?

**Comment by daviesalex** [comment]
@pauldix we are 100% using telegraf on ~1k nodes, with a 10 second collection interval for almost all inputs. Absolutely no writing of old timestamp data that i'm aware of. Those 1,000 nodes seem to be distributed quite nicely over that 10 second period. ~25G on disk data in 24hours.

ls -lah of /data/influxdb-data/metrics/telegraf/default/4 (4 is the only shard on that node) at https://gist.github.com/daviesalex/bc939055f0f5be8920d0

**Comment by pauldix** [comment]
@daviesalex ah, it looks like you're hitting #4388, which I'm in the process of fixing right now. The data files shouldn't be larger than 4GB, which limits the time a full compaction might take. There is a setting you can adjust to have the DB wait longer before it attempts a full compaction: https://github.com/influxdb/influxdb/blob/master/tsdb/config.go#L53

But this one is still a big problem for your setup until it gets fixed. Should have something in the next day or two!

**Comment by daviesalex** [comment]
@pauldix, agreed, thank you - I noticed that massive file as I sent you the ls. At this point we will wait a few days to get #4388 fixed (and/or allow us to WAL a different directory if thats done in this issue), whack everything, and try again.

We will continue to try to do some limited clustering testing, just deleting all the data every 8 hours or so to prevent us hitting this bug. I'll file a separate issue with some questions we already have based on the testing we have done on 0.9.5 so far.

Thank you for your rapid response to these issues!

**Comment by daviesalex** [comment]
We can confirm that #4388 is fixed in 0.9.5-nightly-827c513; our files are now rolling over at about 1.3GB.

We are still seeing these errors although far less frequently:
```
[tcp] 2015/10/23 11:37:43 process write shard error: write shard 4: engine: WAL backed up flushing to index, hit max memory
```

Despite now using PCIe flash for everything (IO service times are ~2ms max, %util never over 3%), but that is a separate issue - we will open a new issue.

Lets leave this issue to track the agreed change (make WAL directory configurable for tsm1).

Thanks!

**Comment by pauldix** [comment]
Fixed in 0.9.6
